test/TransactionSpec.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module TransactionSpec
+  ( transactionTests
+  )
+where
+
+import           Data.Map                       ( Map )
+import qualified Data.Map                      as M
+import           Data.Maybe                     ( maybeToList )
+import           Data.Monoid                    ( (<>) )
+import           Data.Text                      ( Text
+                                                , pack
+                                                )
+import           Test.Hspec
+
+import           Data.Default                   ( Default(..) )
+import           Database.Bolt           hiding ( pack )
+
+mkConnection :: IO Pipe
+mkConnection = connect $ def { user = "neo4j", password = "test" }
+
+anyTxError :: Selector TxError
+anyTxError = const True
+
+transactionTests :: Spec
+transactionTests = describe "Transaction" $ do
+  it "commits" $ do
+    pipe    <- mkConnection
+    persons <- toEntityList "p" <$> run pipe (transact cyphers) :: IO [Person]
+    close pipe
+    persons `shouldBe` input
+  it "rollsback due to malformed cypher query" $ do
+    pipe    <- mkConnection
+    let stmt = toEntityList "g" <$> run pipe (transact failedCyphers) :: IO [Guitar]
+    stmt `shouldThrow` anyTxError
+    close pipe
+
+-- Test input data
+
+input :: [Person]
+input = [Person "Simon" 61, Person "Philip" 63, Person "Erik" 56]
+
+cyphers :: [Cypher]
+cyphers = toCypher <$> input where
+  toCypher p = Cypher
+    (  "CREATE (p:Person { name: '"
+    <> name p
+    <> "', age: "
+    <> pack (show $ age p)
+    <> " } ) RETURN p"
+    )
+    M.empty
+
+failedCyphers :: [Cypher]
+failedCyphers =
+  [ Cypher "CREATE (g:Guitar { brand: 'Gibson' } ) RETURN g"   M.empty
+  , Cypher "CREATE (g:Guitar { brand: 'Fender' } ) RETURN g"   M.empty
+  , Cypher "CREATE (g:Guitar { brand: 'Ibanez' } ) RETURN g"   M.empty
+  , Cypher "This is going to make it fail, BOOM!"              M.empty
+  , Cypher "CREATE (g:Guitar { brand: 'Schecter' } ) RETURN g" M.empty
+  ]
+
+-- Record parser helpers
+
+type NodeProps = Map Text Value
+
+class NodeMapper a where
```

Comment:
These are all the helpers I used in [musikell](https://github.com/gvolpe/musikell/blob/master/src/Repository/Mapper.hs), maybe worth adding them to the library in a future PR?